### DISCUSSION TITLE
Allow for trigger of snippet inside dynamic node from pr #1137: "Allow self-dependent dynamicNode"

### DISCUSSION THREAD

**Discussion by sandersantema** [L3MON4D3/LuaSnip#1164] [Q&A] [answered]
I was looking at #1137 and adjusted the example to my needs:

```lua
  s('test', {
    d(1, function(args)
      if args[1] and args[1][1] and string.match(args[1][1], '[^%w\\]') then
        return sn(nil, { t('{ '), i(1, args[1], { key = 'inner' }), t(' }') })
      else
        return sn(nil, { i(1, args[1], { key = 'inner' }) })
      end
    end, { opt_arg(k('inner')) }),
  }),
```
The purpose of this snippet is to add curly braces when anything other than an optional backslash followed by one or more word characters is typed.

This would be great for latex snippets, but the dynamic node doesn't update anymore after triggering another snippet inside. 
When currently in the snippet having typed `x` and I for example trigger my squared autosnippet `sq` (I really dislike moving my hands too far from the home row :) ) `x^2` should become `{x^2}` as `string.match(args[1][1], '[^%w\\]')` now matches for `^` (and `2`) but instead the snippet stops updating.

**Comment by L3MON4D3** [comment, marked as answer]
Ohh, thanks for reporting that, yeah, there's some bad logic in there, I'll look into it :)

**Reply by L3MON4D3** [reply]
Hey, it should work now, could you update and confirm?

**Reply by sandersantema** [reply]
Almost I believe, these are the autosnippets I used for testing:
```lua
  s('foo', fmta([[\texttt{<>}]], { i(1) })),
  s('test', {
    d(1, function(args)
      if args[1] and args[1][1] and string.match(args[1][1], '[^%w\\]') then
        return sn(nil, { t('{ '), i(1, args[1], { key = 'inner' }), t(' }') })
      else
        return sn(nil, { i(1, args[1], { key = 'inner' }) })
      end
    end, { opt_arg(k('inner')) }),
  }),
```

When `foo` is triggered `{ \texttt{} }` is entered correctly but with `\texttt{}` selected instead of, what I would expect, inside node one of the `foo` snippet with the `test` snippet still working. So, with the cursor represented as `|`, like `{ \texttt{|} }`.
But this might be intended behavior? I think the behavior I expected would make very useful snippets possible however.

**Reply by L3MON4D3** [reply]
Ah, yeah that would be expected.. the issue is that the dynamicNode does not pass through information on what parts of the captured Text is a snippet.
So the snippet you just expanded becomes just Text :/

But it would be really cool if we'd allow specifying where a snippet belongs in the new Text.. we'd have to pass a string with some additional data to the dynamicNode-function, and disallow things like inserting into the string where there is a snippet (inserting is ambiguous where two nodes are adjacent, don't know which should get the char), but simple stuff like `..` should be supportable🤔

There's definitely some more stuff to do :D

**Reply by L3MON4D3** [reply]
Hey, I managed to implement this in #1137, give it a spin :)

You'll have to adapt the snippet a tiny bit, 
```lua
  s('test', {
    t":",
    d(1, function(args)
      if args[1] and string.match(args[1]:str(), '[^%w\\]') then
        return sn(nil, { t('{ '), i(1, args[1], { key = 'inner' }), t(' }') })
      else
        return sn(nil, { i(1, args[1], { key = 'inner' }) })
      end

    end, { opt_arg(k('inner')) }, {snippetstring_args = true}),
    t":"
  })
```
should work

**Reply by sandersantema** [reply]
Excuse the late answer. I'm currently not writing much if any LaTeX, which was my reason for wanting this feature. I've been quite busy lately but I'll try and get around to testing it. If I don't I will in a couple of months as I'll be writing latex again by then. Anyhow thanks a lot for implementing it, with functions like these I wouldn't think it far fetched to say neovim has some of the most powerful snippet functionality there is to get.
